Rover/mode_auto.cpp

```diff
@@ -53,25 +53,42 @@ void ModeAuto::update()
             // start/resume the mission (based on MIS_RESTART parameter)
             mission.start_or_resume();
             waiting_to_start = false;
+
+            // initialise mission change check
+            IGNORE_RETURN(mis_change_detector.check_for_mission_change());
         }
     } else {
+        // check for mission changes
+        if (mis_change_detector.check_for_mission_change()) {
+            // if mission is running restart the current command if it is a waypoint command
+            if ((mission.state() == AP_Mission::MISSION_RUNNING) && (_submode == AutoSubMode::Auto_WP)) {
+                if (mission.restart_current_nav_cmd()) {
+                    gcs().send_text(MAV_SEVERITY_CRITICAL, "Auto mission changed, restarted command");
+                } else {
+                    // failed to restart mission for some reason
+                    gcs().send_text(MAV_SEVERITY_CRITICAL, "Auto mission changed but failed to restart command");
+                }
+            }
+        }
+
         mission.update();
     }
 
     switch (_submode) {
         case Auto_WP:
         {
-            if (!g2.wp_nav.reached_destination()) {
+            // check if we've reached the destination
+            if (!g2.wp_nav.reached_destination() || g2.wp_nav.is_fast_waypoint()) {
                 // update navigation controller
                 navigate_to_waypoint();
             } else {
                 // we have reached the destination so stay here
                 if (rover.is_boat()) {
                     if (!start_loiter()) {
-                        stop_vehicle();
+                        start_stop();
```

Comment:
This a existing issue, but this function name is not very useful. This is the whole method:
```
// start stopping vehicle as quickly as possible
void ModeAuto::start_stop()
{
    _submode = Auto_Stop;
}
```
I would vote to remove the method and just use `    _submode = Auto_Stop;`

Reply:
You're right that it's not very useful but I'd prefer to have that extra layer there if it's not too annoying.  I just like the consistency of callers always using a start_xxx to change submode.